src/common/notifications.cpp

```diff
@@ -0,0 +1,89 @@
+#include <winrt/Windows.Foundation.Collections.h>
+#include <winrt/Windows.UI.Notifications.h>
+#include <winrt/Windows.ApplicationModel.Background.h>
+
+using namespace winrt::Windows::ApplicationModel::Background;
+using winrt::Windows::Data::Xml::Dom::XmlDocument;
+using winrt::Windows::UI::Notifications::ToastNotification;
+using winrt::Windows::UI::Notifications::ToastNotificationManager;
+
+namespace
+{
+    constexpr std::wstring_view TASK_NAME = L"PowerToysBackgroundNotificationsHandler";
+    constexpr std::wstring_view TASK_ENTRYPOINT = L"PowerToysNotifications.BackgroundHandler";
+    constexpr std::wstring_view APPLICATION_ID = L"PowerToys";
+}
+
+void notifications::register_background_toast_handler()
+{
+    try
+    {
+        // Re-request access to clean up from our previous versions
+        BackgroundExecutionManager::RemoveAccess();
+        BackgroundExecutionManager::RequestAccessAsync().get();
+
+        BackgroundTaskBuilder builder;
+        ToastNotificationActionTrigger trigger{ APPLICATION_ID };
+        builder.SetTrigger(trigger);
+        builder.TaskEntryPoint(TASK_ENTRYPOINT);
+        builder.Name(TASK_NAME);
+
+        const auto tasks = BackgroundTaskRegistration::AllTasks();
+        const bool already_registered = std::any_of(begin(tasks), end(tasks), [=](const auto& task) {
+            return task.Value().Name() == TASK_NAME;
+        });
+        if (already_registered)
+        {
+            return;
+        }
+        (void)builder.Register();
+    }
+    catch (...)
+    {
+        // Couldn't register the background task, nothing we can do
+    }
+}
+
+void notifications::show_toast(std::wstring_view message)
+{
+    // The toast won't be actually activated in the background, since it doesn't have any buttons
+    show_toast_background_activated({}, message, {});
+}
+
+void notifications::show_toast_background_activated(std::wstring_view background_handler_id, std::wstring_view message, std::vector<std::wstring_view> button_labels)
+{
+    // DO NOT LOCALIZE any string in this function, because they're XML tags and a subject to
+    // https://docs.microsoft.com/en-us/windows/uwp/design/shell/tiles-and-notifications/toast-xml-schema
+
+    std::wstring toast_xml;
+    toast_xml.reserve(1024);
+    toast_xml += LR"(<?xml version="1.0"?><toast><visual><binding template="ToastGeneric"><text>PowerToys</text><text>)";
+    toast_xml += message;
+    toast_xml += L"</text></binding></visual><actions>";
+
+    for (size_t i = 0; i < size(button_labels); ++i)
+    {
+        toast_xml += LR"(<action activationType="background" arguments=")";
+        toast_xml += std::to_wstring(i); // pass the button ID
+        toast_xml += L';';
+        toast_xml += background_handler_id;
+        toast_xml += LR"(" content=")";
+        toast_xml += button_labels[i];
+        toast_xml += LR"("/>)";
+    }
+    toast_xml += L"</actions></toast>";
+
+    XmlDocument toast_xml_doc;
+    toast_xml_doc.LoadXml(toast_xml);
```

Comment:
LoadXml throws exceptions, what about catching them?

Reply:
We don't catch them, because we can't handle this error correctly. If the API user supplied an invalid `plaintext_message`, she will be greeted with exception instantly. If she's not sure if her data is correct, she'll handle the failed `show_toast_background_activated` function as she pleases. We never know what the failure of input data signifies, so we can never handle it.

Reply:
But we may want to report the error and exit instead of crashing.

Reply:
Yes, the caller can wrap the `show_toast_background_activated` call into try/catch block and report the error somehow. The XML parsing failure is a usually an API user's error, i.e. programmer's mistake, so we can't report it to PT user. If it's not an error, the API user will their own catching logic. Maybe we may want to enable the logic in `unhandled_exception_handler.cpp` for all unhandled exceptions to minimize the amount of crashes?

Reply:
I meant to report the error with telemetry,
Let's add a comment to make it clear that `show_toast_background_activated` throws XmlException.

Currently `unhandled_exception_handler` is only used in debug mode. We should consider using it in release mode.